src/Service/Installer.php

```diff
@@ -37,61 +36,75 @@ final class Installer
     private $event;
 
-    public function __construct(string $name, ?KeyDirectory $keys, IOInterface $io, PackageEvent $event)
+    /** @var \GnuPG|null */
+    private $gpg;
+
+    public function __construct(string $name, IOInterface $io, PackageEvent $event)
     {
         $this->name  = $name;
-        $this->keys  = $keys;
         $this->io    = $io;
         $this->event = $event;
     }
 
-    public function install(FileList $fileList) : void
+    public function install(FileList $fileList, ?KeyDirectory $keys, ?GnuPG $gnuPG) : void
     {
         try {
             $packageVersion = PackageVersion::fromPackageEvent($this->event, $this->name);
         } catch (SomebodyElsesProblem $e) {
             $this->io->write($e->getMessage());
             return;
         }
-
+        $this->keys      = $keys;
+        $this->gpg       = $gnuPG;
         $versionReplacer = new VersionConstraintReplacer($packageVersion);
+        $binDir          = $this->event->getComposer()->getConfig()->get('bin-dir');
+
+        if (!file_exists($binDir)) {
+            mkdir($binDir, 0777, true);
+        }
 
         foreach ($fileList->getList() as $file) {
             $this->io->write(sprintf(
-                '  - Downloading artifact in version %2$s from %1$s',
-                $versionReplacer->replace($file->pharUrl()->toString()),
-                $packageVersion->fullVersion()
+                '  - Downloading artifact from %1$s',
+                $versionReplacer->replace($file->pharUrl()->toString())
             ));
 
-            $pharLocation = $this->downloadPhar($versionReplacer, $file);
+            $downloadLocation = $this->downloadAndVerify($versionReplacer, $file);
+            $installLocation  = new SplFileInfo($binDir . DIRECTORY_SEPARATOR . $file->pharName());
+
+            rename($downloadLocation->getPathname(), $installLocation->getPathname());
+            chmod($installLocation->getPathname(), 0755);
+        }
+    }
+
+    private function downloadAndVerify(VersionConstraintReplacer $versionReplacer, File $file): SplFileInfo
+    {
+        $pharLocation = $this->downloadPhar($versionReplacer, $file);
 
-            if (!$file->signatureUrl()) {
-                $this->io->write('  - No digital signature found! Use this file with care!');
-                continue;
-            }
+        if (!$file->signatureUrl()) {
+            $this->io->write('  - <comment>No digital signature found! Use this file with care!</comment>');
+            return $pharLocation;
+        }
 
-            $signatureLocation = $this->downloadSignature($versionReplacer, $file);
-            $this->verifyPharWithSignature($pharLocation, $signatureLocation);
-            $this->io->write('  - PHAR signature successfully verified');
-            unlink($signatureLocation->getPathname());
+        if (null === $this->gpg) {
+            $this->io->write('  - <comment>No GnuPG found to verify signature! Use this file with care!</comment>');
```

Comment:
And here also a `warning`?

Reply:
The formatting of `<warning>` doesn't fit to well in the output flow so I left it as `<comment>` since it is in a different color it s still pretty recognizable.